### PR TITLE
Add optional bulk flux calculation in ufs-datm

### LINKED ISSUE #243: ensure that coupled model can run in the regression test system on Cheyenne.
## Description
Ufs-weather can be tested on Cheyenne in standalone mode. Now that we have a unified repo, we need to add Cheyenne as a maintained platform for the coupled model also.

## Solution
Regression testing for ufs-weather-model in coupled mode should be added on Cheyenne.

### LINKED ISSUE #265: Add bulk flux option for ufs-datm model
## Description

Users of the DATM application require implementation of an optional bulk flux method for calculating the ao-fluxes.

## Solution

- Additional fields (u10m,v10m,t2m,q2m) exported by the DATM can be used by CMEPS to calculate the aofluxes for MOM6 when the bulk flux method is specified.

- A small modification to the nems.configure will be required to add an attribute describing which method will be used by aoflux.

- Additonal tests need to be added for the bulk flux method. 

## Related to
CMEPS issue [#20](https://github.com/NOAA-EMC/CMEPS/issues/20)
DATM issue [#18](https://github.com/NOAA-EMC/NEMSdatm/issues/18)

### LINKED ISSUE #281: DATM-MOM6-CICE6 is not restart reproducible
## Description

The following three tests have been done:
1. datm_12h_cfsr (The model is run for 12 hours using CFSR forcing.)
2. datm_6h_cfsr (The model is run for 6 hours using CFSR forcing.)
3. datm_restart_cfsr (The model is run from hour 6 to hour 12 using restart files generated from test 2.)

Results from test 1 and test 3 are different at the end of the runs.

### To Reproduce:
What compilers/machines are you seeing this with?
Tests are performed on Hera using Intel compiler.

Give explicit steps to reproduce the behavior.
1. git clone https://github.com/binli2337/ufs-weather-model
2. cd ufs-weather-model
3. git checkout feature/restart_test
4. git submodule update --init --recursive
5. cd tests
6. ./rt.sh -k -l rt.conf1

Output from the tests can be found on Hera at /scratch1/NCEPDEV/stmp2/Bin.Li/FV3_RT/RESTART_TESTS.

### LINKED ISSUE #302: compile flags set for CICE and MOM6 are different when compiled as ufs-datm vs ufs-cpld
## Description

The compiler flags used to compile CICE in ufs-datm are different than when compiled for ufs-cpld. This is the source of the non-reproducibility of the ufs-datm noted in issue #281.

### To Reproduce:

Build ufs-weather either as ufs-cpld or ufs-datm outside of the rt system (ie, a direct build using build.sh). Examine build/CICE-interface/CMakeFiles/cice.dir/flags.make. 

For ufs-cpld:

```Fortran_FLAGS = -g -traceback -FR -convert big_endian -assume byterecl -ftz -align array64byte -xHOST -O2 -fp-model precise -module mod   -qopenmp```

For ufs-datm:

```Fortran_FLAGS = -g -traceback -FR -convert big_endian -assume byterecl -ftz -align array64byte -xHOST -module mod   -qopenmp```

The use of ``-fp-model precise`` is required for restart reproducibility in CICE6. It had been present in the machine macros for all supported platforms. I believe the issue is that when compiled as ufs-cpld, it is getting some set of flags set in the FV3 compile; in the DATM this does not happen.

### PR DESCRIPTION
**Pull request by DeniseWorthen**
## Description

Updates both DATM and CMEPS components to allow optional bulk flux formulation.

### Issue(s) addressed

- Issue #265. The optional bulk flux method is set using the ``ocn_surface_flux_scheme`` variable in the nems.configure. The number of iterations in the aoflux scheme is also set with the configuration variable ``flux_iteration``. This is now set at 2 to replicate the NEMS mediator. 

- Adds two tests (cfsr, gefs) using bulk flux option for C96mx100 case

- CMEPS Issue [#21](https://github.com/NOAA-EMC/CMEPS/issues/21). Adds a nems.config variable to set the restart directory. This allows the coupler history files (when used) to be written to the run directory instead of the RESTART directory. Adds guards (ifPresent, ifSet) to the retrieval of``flds_wiso`` in ``med_phases_aoflux`` to allow removal from nems.configure when un-used. 

- Issue #302. Implements a change in the top-level CMakeLists.txt to ensure that both MOM6 and CICE6 are built with the same compiler flags in the ufs-datm as in ufs-cpld. This resolves Issue #281.

- Issue #243. Adds support for running cpld and datm regression tests on Cheyenne Intel. 

## Testing

Due to the change of ``flux_iteration`` and the bug fix for DATM (PR [#21](https://github.com/NOAA-EMC/NEMSdatm/pull/21) ), all baselines will change. New baselines will be added on all supported machines.

## Dependencies

CMEPS PR [#26](https://github.com/NOAA-EMC/CMEPS/pull/26)
DATM PR [#21](https://github.com/NOAA-EMC/NEMSdatm/pull/21)

Co-authors: 
@DusanJovic-NOAA 
@aerorahul 
@JessicaMeixner-NOAA

### CODE REVIEW

**Comment by DeniseWorthen** [comment]
I am going to revert this to a draft PR until testing by the godas group confirms that the optional bulk method is working correctly.

**Comment by DeniseWorthen** [comment]
Submodules have been updated. This PR is ready for baseline generation.  The new BL date will be develop-20201214

**Comment by DeniseWorthen** [comment]
The PR is updated; the new baseline will be develop-20201215.

File: tests/rt.conf [hunk 1]
```diff
@@ -209,33 +209,34 @@ RUN     | fv3_ccpp_esg_HAFS_v0_hwrf_thompson_debug
...
-#RUN     | cpld_satmedmf                                                                                                       | standard | cheyenne.intel                       | fv3 |
 RUN     | cpld_ca                                                                                                             | standard | hera.intel orion.intel wcoss_dell_p3 | fv3 |
+#RUN     | cpld_control                                                                                                        | standard | cheyenne.intel                       | fv3 |
+#RUN     | cpld_2threads                                                                                                       | standard | cheyenne.intel                       |     |
+#RUN     | cpld_decomp                                                                                                         | standard | cheyenne.intel                       |     |
+#RUN     | cpld_satmedmf                                                                                                       | standard | cheyenne.intel                       | fv3 |
```
Comment (climbfuji) [inline tests/rt.conf:222]: These tests will be turned on in a future PR?

**Review by climbfuji — APPROVED** [review]

File: CMakeLists.txt [hunk 2]
```diff
@@ -157,7 +159,7 @@ target_include_directories(fms PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_D
                                       $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/FMS/fms2_io/include>
                                       $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/FMS/mod>
                                       $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/FMS/mpp/include>)
-target_include_directories(fms INTERFACE 
+target_include_directories(fms INTERFACE
                                          $<INSTALL_INTERFACE:${CMAKE_INSTALL_PREFIX}/include>
                                          $<INSTALL_INTERFACE:${CMAKE_INSTALL_PREFIX}/mod>)
```
Comment (aerorahul) [inline CMakeLists.txt]: Can you please tidy up the spacing here?
Thanks!

File: tests/rt.conf (hunk 1 above)
Comment (DeniseWorthen) [reply tests/rt.conf:222]: Yes, I believe the plan is to support Cheyenne as Tier 1 for the ufs-cpld also. But I think we need to resolve whatever issue is causing the compile jobs to fail. Setting max_builds to 3 worked one time, but not another.

**Review by aerorahul — APPROVED** [review]
looks good to me.
Just one minor comment for your consideration.

File: CICE-interface/CMakeLists.txt [hunk 3]
```diff
@@ -9,7 +9,7 @@ elseif(CMAKE_Fortran_COMPILER_ID MATCHES "Intel")
   set(CMAKE_Fortran_FLAGS "-g -traceback")
   set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -FR -convert big_endian -assume byterecl -ftz -align array64byte -xHOST")
   set(CMAKE_Fortran_FLAGS_RELEASE "-O2 -fp-model precise")
-  set(CMAKE_Fortran_FLAGS_DEBUG "-O0 -check uninit -check bounds -check pointers -fpe0 -check noarg_temp_created -link_mpi=dbg")
```
Comment (junwang-noaa) [inline CICE-interface/CMakeLists.txt:12]: We remove this link_mpi because if it is already specified in Intel.cmake?

**Review by junwang-noaa — COMMENTED** [review]
Denise, so the changes in CMEPS and DATM will only change the coupled and datm test results, there is no impact on uncoupled case, is it correct?

**Comment by DeniseWorthen** [comment]
Yes, the tests have all been passing for the uncoupled case (which is what I expected). For example, cheyenne.intel and gnu are both just using copies of the previous baseline.

File: CICE-interface/CMakeLists.txt (hunk 3 above)
Comment (DeniseWorthen) [reply CICE-interface/CMakeLists.txt:12]: Yes, I believe that is what @aerorahul intended. This fix allowed mom6solo to compile on cheyenne. I did test the cpld_debug on hera w/ this change and it did not affect answers. I think this is also the motivation behind issue #330.

Comment (aerorahul) [reply CICE-interface/CMakeLists.txt:12]: @junwang-noaa 
We should not set that as `CMAKE_Fortran_FLAGS` as is currently set in `cmake/Intel.cmake`
It  should be part of `CMAKE_Fortran_FLAGS_DEBUG`. 
This is raised in issue #330

**Comment by DeniseWorthen** [comment]
This will be ready to merge after the CI tests complete.

**Review by DusanJovic-NOAA — APPROVED** [review]

File: CICE-interface/CMakeLists.txt (hunk 3 above)
Comment (junwang-noaa) [reply CICE-interface/CMakeLists.txt:12]: @aerorahul I am confused about "It should be part of CMAKE_Fortran_FLAGS_DEBUG", the link_mpi was defined in CMAKE_Fortran_FLAGS_DEBUG, but the updated code removes it?

Comment (aerorahul) [reply CICE-interface/CMakeLists.txt:12]: It was hard-wired in the `CMake_Fortran_FLAGS_DEBUG` for MOM6 and CICE6 CMakeLists.txt.  This does not work for machines e.g. wcoss2, cheyenne.

That is not the appropriate place for it.  If it should exist, it should be in `Intel.cmake`.

**Review by junwang-noaa — APPROVED** [review]
Baselines develop-20201215 exist on all platforms, regression tests were run against exactly that baseline on all systems except cheyenne.intel.
On cheyenne.intel the tests were run against 20201214, and this baseline is identical to 20201215 (as per "diff -r develop-20201214 develop-20201215").

**Pull request merged**